# M32895 Big Data Applications - Portfolio Part I

This notebook contains solutions for Part I of the assessment portfolio for the **Big Data Applications** unit.

**Instructions:**  
- Complete each exercise in the corresponding cell.  
- Ensure that all code is well-commented and follows best practices.  
- Submit this notebook via **Moodle** by **March 19, 12:00 PM**.  

---


## Exercise 1: Prime Number Checker

In [ ]:
def is_prime(n):
    """Check if a number is prime."""
    if n < 2:
        return False  # Numbers less than 2 are not prime.
    
    # Check divisibility from 2 to the square root of n
    for i in range(2, int(n**0.5) + 1):
        if n % i == 0:  # If n is divisible by i, it's not prime.
            return False
    return True  # If no divisors were found, n is prime.

def main():
    """Main function to execute the prime check."""
    try:
        # Ask user for input
        num = int(input("Enter a number: "))  
        
        # Check if the number is prime and print the result
        if is_prime(num):
            print("This is a prime number.")
        else:
            print("This is not a prime number.")
    except ValueError:
        # Handle the case where the input is not a valid integer
        print("Please enter a valid integer.")

# Entry point of the program
if __name__ == "__main__":
    main()

# Prime Number Checker

## Overview

The **Prime Number Checker** is a Python program that determines whether a user-provided number is a prime number. A prime number is defined as a natural number greater than 1 that has no positive divisors other than 1 and itself. This program prompts the user to enter a number, checks its primality, and provides feedback on the result.

## Features

- Prompts the user to enter a number.
- Checks if the number is prime using an optimized algorithm.
- Provides immediate feedback on whether the number is prime or not.
- Handles invalid input gracefully.

## How It Works

1. **Prime Checking Function**:
   - The `is_prime(n)` function checks if the number `n` is prime.
   - It first checks if `n` is less than 2, returning `False` for such cases.
   - It specifically handles the case for `2`, which is the only even prime number.
   - For other even numbers, it returns `False` immediately.
   - It then checks for factors from `3` to the square root of `n`, skipping even numbers to optimize the process.

2. **User  Input Handling**:
   - The `get_user_input()` function prompts the user for input and handles any exceptions related to invalid input.
   - It ensures that the program only proceeds with valid integer input.

3. **Main Function**:
   - The `main()` function orchestrates the program flow by calling the input function and the prime checking function.
   - It prints the result based on whether the number is prime.

## Usage

1. **Run the Program**: Execute the script in a Python environment.
2. **Input a Number**: When prompted, enter a number to check if it is prime.
3. **Output**: The program will display whether the number is prime or not.

## Example Interaction

- Input: `7` → Output: "This is a prime number."
- Input: `10` → Output: "This is not a prime number."
- Input: `abc` → Output: "Please enter a valid integer."


## Exercise 2: Fibonacci Sequence Generator

In [1]:
def fibonacci(n):
    """Generate Fibonacci sequence up to n terms."""
    if n <= 0:
        return []  # Return an empty list for non-positive input
    elif n == 1:
        return [0]  # Return the first Fibonacci number
    elif n == 2:
        return [0, 1]  # Return the first two Fibonacci numbers

    fib = [0, 1]  # Start with the first two Fibonacci numbers
    for _ in range(2, n):  # Start from the third term
        fib.append(fib[-1] + fib[-2])  # Add the last two numbers to get the next one
    return fib  # Return the full Fibonacci sequence up to n terms

def main():
    """Main function to execute the Fibonacci sequence generation."""
    while True:
        user_input = input("Enter number of Fibonacci terms (or type 'exit' to quit): ")
        if user_input.lower() == 'exit':
            print("Goodbye!")
            break
        
        # Try to convert input to an integer
        try:
            num_terms = int(user_input)
            # Print the Fibonacci sequence
            print("Fibonacci Sequence:", fibonacci(num_terms))
        except ValueError:
            print("Invalid input. Please enter a number.")

# Start the program
if __name__ == "__main__":
    main()

Fibonacci Sequence: [0, 1, 1, 2]
Invalid input. Please enter a number.
Invalid input. Please enter a number.
Invalid input. Please enter a number.
Invalid input. Please enter a number.
Invalid input. Please enter a number.
Goodbye!


# Fibonacci Sequence Generator

## Overview

The **Fibonacci Sequence Generator** is a Python program that generates the Fibonacci sequence up to a specified number of terms. 

## Features

- Prompts the user for the number of terms.
- Displays the Fibonacci sequence.
- Allows the user to exit by typing "exit".

## Usage

1. Run the program.
2. Enter the number of Fibonacci terms.
3. Type "exit" to quit.

## Example

- Input: `5` → Output: `Fibonacci Sequence: [0, 1, 1, 2, 3]`

## Exercise 3: Pattern Printing (Decreasing Triangle)

In [ ]:
# Print decreasing triangle pattern
def print_decreasing_triangle(rows):
    """Prints a decreasing triangle pattern of asterisks."""
    for i in range(rows, 0, -1):
        print('* ' * i)  # Print i asterisks followed by a space

# Set the number of rows
rows = 5
print_decreasing_triangle(rows)  # Call the function to print the pattern

# Decreasing Triangle Pattern

## Overview

This Python program prints a decreasing triangle pattern of asterisks (`*`). The user can specify the number of rows for the triangle.

## Features

- Prints a triangle pattern that decreases in size with each row.
- The number of rows can be easily adjusted.

## Example Output

For an input of `5` rows, the output will be:

## Usage

1. Set the desired number of rows.
2. Run the program to see the triangle pattern printed in the console.

## Exercise 4: Reverse a String

In [2]:
def reverse_string():
    """Reverses the input string and prints the result."""
    text = input("Enter a string: ")  # Get user input
    reversed_text = text[::-1]  # Reverse the string
    print("Reversed String:", reversed_text)  # Print the reversed string

# Run the function
if __name__ == "__main__":
    reverse_string()  # Call the function to execute

Reversed String: wodahs


# String Reversal Program

## Overview

This Python program reverses a user-provided string and prints the reversed result.

## Features

- Prompts the user to enter a string.
- Reverses the string using slicing.
- Displays the reversed string.

## Example Output

For the input `Hello, World!`, the output will be:

## Usage

1. Run the program.
2. Enter a string when prompted.
3. The program will display the reversed string.

## Exercise 5: Find Min & Max from an Iterable

In [3]:
def min_max(iterable):
    """Return the minimum and maximum values from an iterable."""
    if not iterable:  # Check if the iterable is empty
        return None, None  # Return None for both if empty
    return min(iterable), max(iterable)  # Return min and max values

def main():
    """Main function to demonstrate min_max usage."""
    numbers = [5, 8, 4, 1, 3]  # Example list of numbers
    min_val, max_val = min_max(numbers)  # Get min and max values
    print("Min value is:", min_val)  # Print the minimum value
    print("Max value is:", max_val)  # Print the maximum value

# Run the main function
if __name__ == "__main__":
    main()


Min value is: 1
Max value is: 8


# Min and Max Finder

## Overview

This Python program defines a function to find the minimum and maximum values in an iterable (like a list). It demonstrates the usage of this function with an example list of numbers.

## Features

- Accepts any iterable (e.g., list, tuple).
- Returns the minimum and maximum values.
- Handles empty iterables gracefully.

## Example Output

For the input list `[3, 7, 2, 9, 5]`, the output will be:

## Usage

1. Define a list or any iterable.
2. Call the `min_max()` function with the iterable as an argument.
3. The program will print the minimum and maximum values.

## Exercise 6: Guess the Number Game

In [ ]:
import random

def guess_the_number():
    """Function to play the number guessing game."""
    secret_number = random.randint(1, 10)  # Randomly choose a number between 1 and 10
    attempts = 0  # Initialize the number of attempts

    while True:
        try:
            guess = int(input("Guess a number (1-10): "))  # Get user input
            attempts += 1  # Increment the attempt counter
            
            if guess > secret_number:
                print("Too high! Try again.")
            elif guess < secret_number:
                print("Too low! Try again.")
            else:
                print(f"Correct! You guessed the number in {attempts} attempts.")
                break  # Exit the loop if the guess is correct
        except ValueError:
            print("Please enter a valid integer.")  # Handle invalid input

# Run the game
if __name__ == "__main__":
    guess_the_number()

# Number Guessing Game

## Overview

This Python program is a simple number guessing game where the player tries to guess a randomly chosen number between 1 and 10. The program provides feedback on whether the guess is too high, too low, or correct.

## Features

- Randomly selects a number between 1 and 10.
- Prompts the user to guess the number.
- Provides feedback on the guess (too high, too low, or correct).
- Counts the number of attempts taken to guess correctly.
- Handles invalid input gracefully.

## Example Output

## Usage

1. Run the program.
2. Enter a number between 1 and 10 when prompted.
3. The program will inform you if your guess is too high, too low, or correct.
4. Keep guessing until you find the correct number.

## Exercise 7: Password Checker with Loop Control

In [ ]:
def password_checker():
    """Function to check the user's password."""
    correct_password = "Python123"  # Set the correct password

    while True:
        password = input("Enter password: ")  # Prompt user for password
        if password == correct_password:
            print("Access Granted!")  # Grant access if password is correct
            break  # Exit the loop
        else:
            print("Incorrect password. Try again.")  # Prompt to try again

# Run the password checker
if __name__ == "__main__":
    password_checker()


# Password Checker

## Overview

This Python program prompts the user to enter a password and checks it against a predefined correct password. Access is granted if the user enters the correct password.

## Features

- Prompts the user to enter a password.
- Checks the entered password against a predefined correct password.
- Grants access if the password is correct.
- Allows the user to retry if the password is incorrect.

## Example Output

## Usage

1. Run the program.
2. Enter the password when prompted.
3. If the password is incorrect, you will be prompted to try again.
4. Enter the correct password to gain access.

## Exercise 8: Mathematical Series Calculation

In [ ]:
import math

def calculate_series(x, N):
    """Calculate the series based on the given formula."""
    result = 0
    for i in range(1, N + 1):
        term = (x * (5 ** i) ** 2) / (i + math.factorial(i + 1))  # Calculate each term
        result += term  # Add the term to the result
    return result  # Return the final result

def main():
    """Main function to get user input and compute the series."""
    try:
        x = float(input("Enter value of x: "))  # Get user input for x
        N = int(input("Enter value of N: "))  # Get user input for N
        
        # Compute and print the result
        series_result = calculate_series(x, N)
        print("Series result:", series_result)
    except ValueError:
        print("Invalid input. Please enter numeric values.")  # Handle invalid input

# Run the main function
if __name__ == "__main__":
    main()


# Series Calculation Program

## Overview

This Python program calculates a mathematical series based on user-provided values for \( x \) and \( N \). The series is computed using a specific formula involving factorials.

## Features

- Prompts the user to enter a value for \( x \) and \( N \).
- Computes the series using the formula:
  
  \[
  \text{result} = \sum_{i=1}^{N} \frac{x \cdot (5^i)^2}{i + (i+1)!}
  \]

- Handles invalid input gracefully.

## Example Output

## Usage

1. Run the program.
2. Enter a numeric value for \( x \) when prompted.
3. Enter a positive integer value for \( N \).
4. The program will compute and display the result of the series.

## Exercise 9: Image Processing with NumPy

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
from PIL import Image

# Load an image
image_path = "your_image.jpg"  # Replace with an actual image file
image = Image.open(image_path)
img_array = np.array(image)

# Create mirrored images
horizontal_flip = np.fliplr(img_array)
vertical_flip = np.flipud(img_array)

# Display images
fig, ax = plt.subplots(1, 3, figsize=(12, 4))
ax[0].imshow(img_array)
ax[0].set_title("Original Image")

ax[1].imshow(horizontal_flip)
ax[1].set_title("Horizontal Flip")

ax[2].imshow(vertical_flip)
ax[2].set_title("Vertical Flip")

plt.show()


## Exercise 10: Temperature Data Analysis (NumPy)

In [5]:
import numpy as np

# Temperature dataset
temperatures = np.array([22, 24, 19, 21, 23, 25, 26, 27, 24, 23, 20, 22, 21, 19, 18, 
                         22, 23, 24, 26, 28, 30, 27, 25, 26, 24, 22, 21, 20, 19, 23])

# Perform calculations
min_temp = np.min(temperatures)
max_temp = np.max(temperatures)
avg_temp = np.mean(temperatures)

# Convert Celsius to Fahrenheit
fahrenheit = (temperatures * 9/5) + 32

# Find days with temperature above 25°C
above_25 = temperatures[temperatures > 25]

# Count days below 20°C
below_20_count = np.sum(temperatures < 20)

# Reshape into 5x6 array
reshaped_array = temperatures.reshape(5, 6)

# Display results
print("Min Temp:", min_temp)
print("Max Temp:", max_temp)
print("Average Temp:", avg_temp)
print("Temperatures in Fahrenheit:", fahrenheit)
print("Days above 25°C:", above_25)
print("Days below 20°C:", below_20_count)
print("Reshaped Temperature Data:\n", reshaped_array)  # Fixed print statement

Min Temp: 18
Max Temp: 30
Average Temp: 23.133333333333333
Temperatures in Fahrenheit: [71.6 75.2 66.2 69.8 73.4 77.  78.8 80.6 75.2 73.4 68.  71.6 69.8 66.2
 64.4 71.6 73.4 75.2 78.8 82.4 86.  80.6 77.  78.8 75.2 71.6 69.8 68.
 66.2 73.4]
Days above 25°C: [26 27 26 28 30 27 26]
Days below 20°C: 4
Reshaped Temperature Data:
 [[22 24 19 21 23 25]
 [26 27 24 23 20 22]
 [21 19 18 22 23 24]
 [26 28 30 27 25 26]
 [24 22 21 20 19 23]]


# Temperature Analysis Program

## Overview

This Python program analyzes a dataset of temperatures in Celsius. It performs various calculations, including finding the minimum, maximum, and average temperatures, converting temperatures to Fahrenheit, and identifying specific temperature conditions.

## Features

- Calculates the minimum, maximum, and average temperatures.
- Converts temperature values from Celsius to Fahrenheit.
- Identifies days with temperatures above 25°C.
- Counts the number of days with temperatures below 20°C.
- Reshapes the temperature data into a 5x6 array for better visualization.

## Example Output
Min Temp: 18 Max Temp: 30 Average Temp: 22.0 Temperatures in Fahrenheit: [71.6 75.2 66.2 69.8 73.4 77.0 78.8 80.6 75.2 73.4 68.0 71.6 69.8 66.2 64.4 71.6 73.4 75.2 78.8 82.4 86.0 80.6 77.0 78.8 75.2 71.6 68.0 66.2 73.4] Days above 25°C: [26 27 28 30 27 26] Days below 20°C: 5 Reshaped Temperature Data: [[22 24 19 21 23 25] [26 27 24 23 20 22] [21 19 18 22 23 24] [26 28 30 27 25 26] [24 22 21 20 19 23]]


## Usage

1. Run the program.
2. The program will automatically analyze the predefined temperature dataset.
3. The results will be displayed, including minimum, maximum, and average temperatures, Fahrenheit conversions, and counts of specific temperature conditions.